File: C/actors.c
// Copyright (c) 1980, InfoCom Computers and Communications, Cambridge MA 02142
// All rights reserved, commercial usage strictly prohibited.
// Written by R. M. Supnik.
// Revisions Copyright (c) 2021, Darth Spectra (Lydia Marie Williamson).
#include "F2C.h"
#include "extern.h"
#include "common.h"

// Applicables for adventurers
Bool aappli(int ri) {
// System generated locals
   Bool ret_val;

// Local variables
   Bool f;
   int i;

// AAPPLI, PAGE 2

   if (ri == 0) {
      goto L10;
   }
// 						!IF ZERO, NO APP.
   ret_val = true;
// 						!ASSUME WINS.
   switch (ri) {
      case 1:
         goto L1000;
      case 2:
         goto L2000;
   }
// 						!BRANCH ON ADV.
   bug(11, ri);

// COMMON FALSE RETURN.

L10:
   ret_val = false;
   return ret_val;

// A1--	ROBOT.  PROCESS MOST COMMANDS GIVEN TO ROBOT.

L1000:
   if (prsvec_1.prsa != vindex_1.raisew || prsvec_1.prso != oindex_1.rcage) {
      goto L1200;
   }
   cevent_1.cflag[cindex_1.cevsph - 1] = false;
// 						!ROBOT RAISED CAGE.
   play_1.winner = aindex_1.player;
// 						!RESET FOR PLAYER.
   f = moveto(rindex_1.cager, play_1.winner);
// 						!MOVE TO NEW ROOM.
   newsta(oindex_1.cage, 567, rindex_1.cager, 0, 0);
// 						!INSTALL CAGE IN ROOM.
   newsta(oindex_1.robot, 0, rindex_1.cager, 0, 0);
// 						!INSTALL ROBOT IN ROOM.
   advs_1.aroom[aindex_1.arobot - 1] = rindex_1.cager;
// 						!ALSO MOVE ROBOT/ADV.
   findex_1.cagesf = true;
// 						!CAGE SOLVED.
   objcts_1.oflag1[oindex_1.robot - 1] &= ~NDscO;
   objcts_1.oflag1[oindex_1.spher - 1] |= TakeO;
   return ret_val;

L1200:
   if (prsvec_1.prsa != vindex_1.drinkw && prsvec_1.prsa != vindex_1.eatw) {
      goto L1300;
   }
   rspeak(568);
// 						!EAT OR DRINK, JOKE.
   return ret_val;

L1300:
   if (prsvec_1.prsa != vindex_1.readw) {
      goto L1400;
   }
// 						!READ,
   rspeak(569);
// 						!JOKE.
   return ret_val;

L1400:
   if (prsvec_1.prsa == vindex_1.walkw || prsvec_1.prsa == vindex_1.takew || prsvec_1.prsa == vindex_1.dropw || prsvec_1.prsa == vindex_1.putw || prsvec_1.prsa == vindex_1.pushw || prsvec_1.prsa == vindex_1.throww || prsvec_1.prsa == vindex_1.turnw || prsvec_1.prsa == vindex_1.leapw) {
      goto L10;
   }
   rspeak(570);
// 						!JOKE.
   return ret_val;
// AAPPLI, PAGE 3

// A2--	MASTER.  PROCESS MOST COMMANDS GIVEN TO MASTER.

L2000:
   if ((objcts_1.oflag2[oindex_1.qdoor - 1] & OpenO) != 0) {
      goto L2100;
   }
   rspeak(783);
// 						!NO MASTER YET.
   return ret_val;

L2100:
   if (prsvec_1.prsa != vindex_1.walkw) {
      goto L2200;
   }
// 						!WALK?
   i = 784;
// 						!ASSUME WONT.
   if (play_1.here == rindex_1.scorr && (prsvec_1.prso == xsrch_1.xnorth || prsvec_1.prso == xsrch_1.xenter) || play_1.here == rindex_1.ncorr && (prsvec_1.prso == xsrch_1.xsouth || prsvec_1.prso == xsrch_1.xenter)) {
      i = 785;
   }
   rspeak(i);
   return ret_val;

L2200:
   if (prsvec_1.prsa == vindex_1.takew || prsvec_1.prsa == vindex_1.dropw || prsvec_1.prsa == vindex_1.putw || prsvec_1.prsa == vindex_1.throww || prsvec_1.prsa == vindex_1.pushw || prsvec_1.prsa == vindex_1.turnw || prsvec_1.prsa == vindex_1.spinw || prsvec_1.prsa == vindex_1.trntow || prsvec_1.prsa == vindex_1.follow || prsvec_1.prsa == vindex_1.stayw || prsvec_1.prsa == vindex_1.openw || prsvec_1.prsa == vindex_1.closew || prsvec_1.prsa == vindex_1.killw) {
      goto L10;
   }
   rspeak(786);
// 						!MASTER CANT DO IT.
   return ret_val;
}

// Intermove thief demon
void thiefd(void) {
// System generated locals
   int i__1, i__2;

// Local variables
   int i, j, nr;
   Bool once;
   int rhere;

// THIEFD, PAGE 2

#ifdef ALLOW_GDT
   dflag = (debug_1.prsflg & 32768) != 0;
#endif
// 						!SET UP DETAIL FLAG.
   once = false;
// 						!INIT FLAG.
L1025:
   rhere = objcts_1.oroom[oindex_1.thief - 1];
// 						!VISIBLE POS.
   if (rhere != 0) {
      hack_1.thfpos = rhere;
   }

   if (hack_1.thfpos == play_1.here) {
      goto L1100;
   }
// 						!THIEF IN WIN RM?
   if (hack_1.thfpos != rindex_1.treas) {
      goto L1400;
   }
// 						!THIEF NOT IN TREAS?

// THIEF IS IN TREASURE ROOM, AND WINNER IS NOT.

#ifdef ALLOW_GDT
   if (dflag) print(" THIEFD-- IN TREASURE ROOM");
#endif
   if (rhere == 0) {
      goto L1050;
   }
// 						!VISIBLE?
   newsta(oindex_1.thief, 0, 0, 0, 0);
// 						!YES, VANISH.
   rhere = 0;
   if (qhere(oindex_1.still, rindex_1.treas) || objcts_1.oadv[oindex_1.still - 1] == -oindex_1.thief) {
      newsta(oindex_1.still, 0, 0, oindex_1.thief, 0);
   }
L1050:
   i__1 = -oindex_1.thief;
   i = robadv(i__1, hack_1.thfpos, 0, 0);
// 						!DROP VALUABLES.
   if (qhere(oindex_1.egg, hack_1.thfpos)) {
      objcts_1.oflag2[oindex_1.egg - 1] |= OpenO;
   }
   goto L1700;

// THIEF AND WINNER IN SAME ROOM.

L1100:
   if (hack_1.thfpos == rindex_1.treas) {
      goto L1700;
   }
// 						!IF TREAS ROOM, NOTHING.
   if ((rooms_1.rflag[hack_1.thfpos - 1] & LightR) != 0) {
      goto L1400;
   }
#ifdef ALLOW_GDT
   if (dflag) print(" THIEFD-- IN ADV ROOM");
#endif
   if (hack_1.thfflg) {
      goto L1300;
   }
// 						!THIEF ANNOUNCED?
   if (rhere != 0 || prob(70, 70)) {
      goto L1150;
   }
// 						!IF INVIS AND 30%.
   if (objcts_1.ocan[oindex_1.still - 1] != oindex_1.thief) {
      goto L1700;
   }
// 						!ABORT IF NO STILLETTO.
   newsta(oindex_1.thief, 583, hack_1.thfpos, 0, 0);
// 						!INSERT THIEF INTO ROOM.
   hack_1.thfflg = true;
// 						!THIEF IS ANNOUNCED.
   return;

L1150:
   if (rhere == 0 || (objcts_1.oflag2[oindex_1.thief - 1] & FiteO) == 0) {
      goto L1200;
   }
   if (winnin(oindex_1.thief, play_1.winner)) {
      goto L1175;
   }
// 						!WINNING?
   newsta(oindex_1.thief, 584, 0, 0, 0);
// 						!NO, VANISH THIEF.
   objcts_1.oflag2[oindex_1.thief - 1] &= ~FiteO;
   if (qhere(oindex_1.still, hack_1.thfpos) || objcts_1.oadv[oindex_1.still - 1] == -oindex_1.thief) {
      newsta(oindex_1.still, 0, 0, oindex_1.thief, 0);
   }
   return;

L1175:
   if (prob(90, 90)) {
      goto L1700;
   }
// 						!90% CHANCE TO STAY.

L1200:
   if (rhere == 0 || prob(70, 70)) {
      goto L1250;
   }
// 						!IF VISIBLE AND 30%
   newsta(oindex_1.thief, 585, 0, 0, 0);
// 						!VANISH THIEF.
   if (qhere(oindex_1.still, hack_1.thfpos) || objcts_1.oadv[oindex_1.still - 1] == -oindex_1.thief) {
      newsta(oindex_1.still, 0, 0, oindex_1.thief, 0);
   }
   return;

L1300:
   if (rhere == 0) {
      goto L1700;
   }
// 						!ANNOUNCED.  VISIBLE?
L1250:
   if (prob(70, 70)) {
      return;
   }
// 						!70% CHANCE TO DO NOTHING.
   hack_1.thfflg = true;
   i__1 = -oindex_1.thief;
   i__2 = -oindex_1.thief;
   nr = robrm(hack_1.thfpos, 100, 0, 0, i__1) + robadv(play_1.winner, 0, 0, i__2);
   i = 586;
// 						!ROBBED EM.
   if (rhere != 0) {
      i = 588;
   }
// 						!WAS HE VISIBLE?
   if (nr != 0) {
      ++i;
   }
// 						!DID HE GET ANYTHING?
   newsta(oindex_1.thief, i, 0, 0, 0);
// 						!VANISH THIEF.
   if (qhere(oindex_1.still, hack_1.thfpos) || objcts_1.oadv[oindex_1.still - 1] == -oindex_1.thief) {
      newsta(oindex_1.still, 0, 0, oindex_1.thief, 0);
   }
   if (nr != 0 && !lit(hack_1.thfpos)) {
      rspeak(406);
   }
   rhere = 0;
   goto L1700;
// 						!ONWARD.

// NOT IN ADVENTURERS ROOM.

L1400:
   newsta(oindex_1.thief, 0, 0, 0, 0);
// 						!VANISH.
   rhere = 0;
#ifdef ALLOW_GDT
   if (dflag) print(" THIEFD-- IN ROOM %I4", hack_1.thfpos);
#endif
   if (qhere(oindex_1.still, hack_1.thfpos) || objcts_1.oadv[oindex_1.still - 1] == -oindex_1.thief) {
      newsta(oindex_1.still, 0, 0, oindex_1.thief, 0);
   }
   if ((rooms_1.rflag[hack_1.thfpos - 1] & SeenR) == 0) {
      goto L1700;
   }
   i__1 = -oindex_1.thief;
   i = robrm(hack_1.thfpos, 75, 0, 0, i__1);
// 						!ROB ROOM 75%.
   if (hack_1.thfpos < rindex_1.maze1 || hack_1.thfpos > rindex_1.maz15 || play_1.here < rindex_1.maze1 || play_1.here > rindex_1.maz15) {
      goto L1500;
   }
   i__1 = objcts_1.olnt;
   for (i = 1; i <= i__1; ++i) {
// 						!BOTH IN MAZE.
      if (!qhere(i, hack_1.thfpos) || prob(60, 60) || (objcts_1.oflag1[i - 1] & VisiO + TakeO) != VisiO + TakeO) {
         goto L1450;
      }
      rspsub(590, objcts_1.odesc2[i - 1]);
// 						!TAKE OBJECT.
      if (prob(40, 20)) {
         goto L1700;
      }
      i__2 = -oindex_1.thief;
      newsta(i, 0, 0, 0, i__2);
// 						!MOST OF THE TIME.
      objcts_1.oflag2[i - 1] |= TChO;
      goto L1700;
   L1450:
      ;
   }
   goto L1700;

L1500:
   i__1 = objcts_1.olnt;
   for (i = 1; i <= i__1; ++i) {
// 						!NOT IN MAZE.
      if (!qhere(i, hack_1.thfpos) || objcts_1.otval[i - 1] != 0 || prob(80, 60) || (objcts_1.oflag1[i - 1] & VisiO + TakeO) != VisiO + TakeO) {
         goto L1550;
      }
      i__2 = -oindex_1.thief;
      newsta(i, 0, 0, 0, i__2);
      objcts_1.oflag2[i - 1] |= TChO;
      goto L1700;
   L1550:
      ;
   }

// NOW MOVE TO NEW ROOM.

L1700:
   if (objcts_1.oadv[oindex_1.rope - 1] == -oindex_1.thief) {
      findex_1.domef = false;
   }
   if (once) {
      goto L1800;
   }
   once = !once;
L1750:
   --hack_1.thfpos;
// 						!NEXT ROOM.
   if (hack_1.thfpos <= 0) {
      hack_1.thfpos = rooms_1.rlnt;
   }
   if ((rooms_1.rflag[hack_1.thfpos - 1] & LandR + SacrdR + EndR) != LandR) {
      goto L1750;
   }
   hack_1.thfflg = false;
// 						!NOT ANNOUNCED.
   goto L1025;
// 						!ONCE MORE.

// ALL DONE.

L1800:
   if (hack_1.thfpos == rindex_1.treas) {
      return;
   }
// 						!IN TREASURE ROOM?
   j = 591;
// 						!NO, DROP STUFF.
   if (hack_1.thfpos != play_1.here) {
      j = 0;
   }
   i__1 = objcts_1.olnt;
   for (i = 1; i <= i__1; ++i) {
      if (objcts_1.oadv[i - 1] != -oindex_1.thief || prob(70, 70) || objcts_1.otval[i - 1] > 0) {
         goto L1850;
      }
      newsta(i, j, hack_1.thfpos, 0, 0);
      j = 0;
   L1850:
      ;
   }
}

File: actors.f
C Copyright (c) 1980, InfoCom Computers and Communications, Cambridge MA 02142
C All rights reserved, commercial usage strictly prohibited.
C Written by R. M. Supnik.
C Revisions Copyright (c) 2021, Darth Spectra (Lydia Marie Williamson).
C
C AAPPLI- APPLICABLES FOR ADVENTURERS
	LOGICAL FUNCTION AAPPLI(RI)
	IMPLICIT INTEGER (A-Z)
	LOGICAL F,MOVETO
	include 'parser.h'
	include 'gamestat.h'
	include 'rooms.h'
	include 'rflag.h'
	include 'rindex.h'
	include 'xsrch.h'
	include 'objects.h'
	include 'oflags.h'
	include 'oindex.h'
	include 'clock.h'
	include 'advers.h'
	include 'verbs.h'
	include 'flags.h'
C AAPPLI, PAGE 2
C
	IF(RI.EQ.0) GO TO 10
C						!IF ZERO, NO APP.
	AAPPLI=.TRUE.
C						!ASSUME WINS.
	GO TO (1000,2000),RI
C						!BRANCH ON ADV.
	CALL BUG(11,RI)
C
C COMMON FALSE RETURN.
C
10	AAPPLI=.FALSE.
	RETURN
C
C A1--	ROBOT.  PROCESS MOST COMMANDS GIVEN TO ROBOT.
C
1000	IF((PRSA.NE.RAISEW).OR.(PRSO.NE.RCAGE)) GO TO 1200
	CFLAG(CEVSPH)=.FALSE.
C						!ROBOT RAISED CAGE.
	WINNER=PLAYER
C						!RESET FOR PLAYER.
	F=MOVETO(CAGER,WINNER)
C						!MOVE TO NEW ROOM.
	CALL NEWSTA(CAGE,567,CAGER,0,0)
C						!INSTALL CAGE IN ROOM.
	CALL NEWSTA(ROBOT,0,CAGER,0,0)
C						!INSTALL ROBOT IN ROOM.
	AROOM(AROBOT)=CAGER
C						!ALSO MOVE ROBOT/ADV.
	CAGESF=.TRUE.
C						!CAGE SOLVED.
	OFLAG1(ROBOT)=IAND(OFLAG1(ROBOT),NOT(NDSCBT))
	OFLAG1(SPHER)=IOR(OFLAG1(SPHER),TAKEBT)
	RETURN
C
1200	IF((PRSA.NE.DRINKW).AND.(PRSA.NE.EATW)) GO TO 1300
	CALL RSPEAK(568)
C						!EAT OR DRINK, JOKE.
	RETURN
C
1300	IF(PRSA.NE.READW) GO TO 1400
C						!READ,
	CALL RSPEAK(569)
C						!JOKE.
	RETURN
C
1400	IF((PRSA.EQ.WALKW).OR.(PRSA.EQ.TAKEW).OR.(PRSA.EQ.DROPW)
     & .OR.(PRSA.EQ.PUTW).OR.(PRSA.EQ.PUSHW).OR.(PRSA.EQ.THROWW)
     & .OR.(PRSA.EQ.TURNW).OR.(PRSA.EQ.LEAPW)) GO TO 10
	CALL RSPEAK(570)
C						!JOKE.
	RETURN
C AAPPLI, PAGE 3
C
C A2--	MASTER.  PROCESS MOST COMMANDS GIVEN TO MASTER.
C
2000	IF(IAND(OFLAG2(QDOOR),OPENBT).NE.0) GO TO 2100
	CALL RSPEAK(783)
C						!NO MASTER YET.
	RETURN
C
2100	IF(PRSA.NE.WALKW) GO TO 2200
C						!WALK?
	I=784
C						!ASSUME WONT.
	IF(((HERE.EQ.SCORR).AND.
     &	((PRSO.EQ.XNORTH).OR.(PRSO.EQ.XENTER))).OR.
     &  ((HERE.EQ.NCORR).AND.
     &	((PRSO.EQ.XSOUTH).OR.(PRSO.EQ.XENTER))))
     &	I=785
	CALL RSPEAK(I)
	RETURN
C
2200	IF((PRSA.EQ.TAKEW).OR.(PRSA.EQ.DROPW).OR.(PRSA.EQ.PUTW).OR.
     &  (PRSA.EQ.THROWW).OR.(PRSA.EQ.PUSHW).OR.(PRSA.EQ.TURNW).OR.
     &  (PRSA.EQ.SPINW).OR.(PRSA.EQ.TRNTOW).OR.(PRSA.EQ.FOLLOW).OR.
     &  (PRSA.EQ.STAYW).OR.(PRSA.EQ.OPENW).OR.(PRSA.EQ.CLOSEW).OR.
     &  (PRSA.EQ.KILLW)) GO TO 10
	CALL RSPEAK(786)
C						!MASTER CANT DO IT.
	RETURN
	END
C
C THIEFD-	INTERMOVE THIEF DEMON
	SUBROUTINE THIEFD
	IMPLICIT INTEGER (A-Z)
	LOGICAL ONCE,PROB,QHERE,QSTILL,LIT,WINNIN
	INCLUDE 'gamestat.h'
C
	INCLUDE 'debug.h'
	INCLUDE 'rooms.h'
	INCLUDE 'rflag.h'
	INCLUDE 'rindex.h'
	INCLUDE 'objects.h'
	INCLUDE 'oflags.h'
	INCLUDE 'oindex.h'
	INCLUDE 'villians.h'
	INCLUDE 'flags.h'
C
C In-line function
	QSTILL(R)=(QHERE(STILL,R).OR.(OADV(STILL).EQ.-THIEF))
C THIEFD, PAGE 2
C
D	DFLAG=IAND(PRSFLG, 32768).NE.0
C						!SET UP DETAIL FLAG.
	ONCE=.FALSE.
C						!INIT FLAG.
1025	RHERE=OROOM(THIEF)
C						!VISIBLE POS.
	IF(RHERE.NE.0) THFPOS=RHERE
C
	IF(THFPOS.EQ.HERE) GO TO 1100
C						!THIEF IN WIN RM?
	IF(THFPOS.NE.TREAS) GO TO 1400
C						!THIEF NOT IN TREAS?
C
C THIEF IS IN TREASURE ROOM, AND WINNER IS NOT.
C
D	IF(DFLAG) PRINT 10
D10	FORMAT(' THIEFD-- IN TREASURE ROOM')
	IF(RHERE.EQ.0) GO TO 1050
C						!VISIBLE?
	CALL NEWSTA(THIEF,0,0,0,0)
C						!YES, VANISH.
	RHERE=0
	IF(QSTILL(TREAS)) CALL NEWSTA(STILL,0,0,THIEF,0)
1050	I=ROBADV(-THIEF,THFPOS,0,0)
C						!DROP VALUABLES.
	IF(QHERE(EGG,THFPOS)) OFLAG2(EGG)=IOR(OFLAG2(EGG),OPENBT)
	GO TO 1700
C
C THIEF AND WINNER IN SAME ROOM.
C
1100	IF(THFPOS.EQ.TREAS) GO TO 1700
C						!IF TREAS ROOM, NOTHING.
	IF(IAND(RFLAG(THFPOS),RLIGHT).NE.0) GO TO 1400
D	IF(DFLAG) PRINT 20
D20	FORMAT(' THIEFD-- IN ADV ROOM')
	IF(THFFLG) GO TO 1300
C						!THIEF ANNOUNCED?
	IF((RHERE.NE.0).OR.PROB(70,70)) GO TO 1150
C						!IF INVIS AND 30%.
	IF(OCAN(STILL).NE.THIEF) GO TO 1700
C						!ABORT IF NO STILLETTO.
	CALL NEWSTA(THIEF,583,THFPOS,0,0)
C						!INSERT THIEF INTO ROOM.
	THFFLG=.TRUE.
C						!THIEF IS ANNOUNCED.
	RETURN
C
1150	IF((RHERE.EQ.0).OR.(IAND(OFLAG2(THIEF),FITEBT).EQ.0))
     &	GO TO 1200
	IF(WINNIN(THIEF,WINNER)) GO TO 1175
C						!WINNING?
	CALL NEWSTA(THIEF,584,0,0,0)
C						!NO, VANISH THIEF.
	OFLAG2(THIEF)=IAND(OFLAG2(THIEF), not(FITEBT))
	IF(QSTILL(THFPOS)) CALL NEWSTA(STILL,0,0,THIEF,0)
	RETURN
C
1175	IF(PROB(90,90)) GO TO 1700
C						!90% CHANCE TO STAY.
C
1200	IF((RHERE.EQ.0).OR.PROB(70,70)) GO TO 1250
C						!IF VISIBLE AND 30%
	CALL NEWSTA(THIEF,585,0,0,0)
C						!VANISH THIEF.
	IF(QSTILL(THFPOS)) CALL NEWSTA(STILL,0,0,THIEF,0)
	RETURN
C
1300	IF(RHERE.EQ.0) GO TO 1700
C						!ANNOUNCED.  VISIBLE?
1250	IF(PROB(70,70)) RETURN
C						!70% CHANCE TO DO NOTHING.
	THFFLG=.TRUE.
	NR=ROBRM(THFPOS,100,0,0,-THIEF)+ROBADV(WINNER,0,0,-THIEF)
	I=586
C						!ROBBED EM.
	IF(RHERE.NE.0) I=588
C						!WAS HE VISIBLE?
	IF(NR.NE.0) I=I+1
C						!DID HE GET ANYTHING?
	CALL NEWSTA(THIEF,I,0,0,0)
C						!VANISH THIEF.
	IF(QSTILL(THFPOS))
     &	CALL NEWSTA(STILL,0,0,THIEF,0)
	IF((NR.NE.0).AND..NOT.LIT(THFPOS)) CALL RSPEAK(406)
	RHERE=0
	GO TO 1700
C						!ONWARD.
C
C NOT IN ADVENTURERS ROOM.
C
1400	CALL NEWSTA(THIEF,0,0,0,0)
C						!VANISH.
	RHERE=0
D	IF(DFLAG) PRINT 30,THFPOS
D30	FORMAT(' THIEFD-- IN ROOM ',I4)
	IF(QSTILL(THFPOS))
     &	CALL NEWSTA(STILL,0,0,THIEF,0)
	IF(IAND(RFLAG(THFPOS),RSEEN).EQ.0) GO TO 1700
	I=ROBRM(THFPOS,75,0,0,-THIEF)
C						!ROB ROOM 75%.
	IF((THFPOS.LT.MAZE1).OR.(THFPOS.GT.MAZ15).OR.
     &	(HERE.LT.MAZE1).OR.(HERE.GT.MAZ15)) GO TO 1500
	DO 1450 I=1,OLNT
C						!BOTH IN MAZE.
	  IF(.NOT.QHERE(I,THFPOS).OR.PROB(60,60).OR.
     &	(IAND(OFLAG1(I),(VISIBT+TAKEBT)).NE.(VISIBT+TAKEBT)))
     &	GO TO 1450
	  CALL RSPSUB(590,ODESC2(I))
C						!TAKE OBJECT.
	  IF(PROB(40,20)) GO TO 1700
	  CALL NEWSTA(I,0,0,0,-THIEF)
C						!MOST OF THE TIME.
	  OFLAG2(I)=IOR(OFLAG2(I),TCHBT)
	  GO TO 1700
1450	CONTINUE
	GO TO 1700
C
1500	DO 1550 I=1,OLNT
C						!NOT IN MAZE.
	  IF(.NOT.QHERE(I,THFPOS).OR.(OTVAL(I).NE.0).OR.PROB(80,60).OR.
     &	(IAND(OFLAG1(I),(VISIBT+TAKEBT)).NE.(VISIBT+TAKEBT)))
     &	GO TO 1550
	  CALL NEWSTA(I,0,0,0,-THIEF)
	  OFLAG2(I)=IOR(OFLAG2(I),TCHBT)
	  GO TO 1700
1550	CONTINUE
C
C NOW MOVE TO NEW ROOM.
C
1700	IF(OADV(ROPE).EQ.-THIEF) DOMEF=.FALSE.
	IF(ONCE) GO TO 1800
	ONCE=.NOT.ONCE
1750	THFPOS=THFPOS-1
C						!NEXT ROOM.
	IF(THFPOS.LE.0) THFPOS=RLNT
	IF(IAND(RFLAG(THFPOS),(RLAND+RSACRD+REND)).NE.RLAND)
     &	GO TO 1750
	THFFLG=.FALSE.
C						!NOT ANNOUNCED.
	GO TO 1025
C						!ONCE MORE.
C
C ALL DONE.
C
1800	IF(THFPOS.EQ.TREAS) RETURN
C						!IN TREASURE ROOM?
	J=591
C						!NO, DROP STUFF.
	IF(THFPOS.NE.HERE) J=0
	DO 1850 I=1,OLNT
	  IF((OADV(I).NE.-THIEF).OR.PROB(70,70).OR.
     &	(OTVAL(I).GT.0)) GO TO 1850
	  CALL NEWSTA(I,J,THFPOS,0,0)
	  J=0
1850	CONTINUE
	RETURN
	END
